qrmi/dependencies/pasqal_cloud_client/src/client.rs

```diff
@@ -0,0 +1,263 @@
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+//! Pasqal Cloud API Client
+
+use anyhow::{bail, Result};
+
+use crate::models::batch::BatchStatus;
+use crate::models::device::DeviceType;
+use log::info;
+use reqwest::header;
+use reqwest_middleware::ClientBuilder as ReqwestClientBuilder;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// An asynchronous `Client` to make Requests with.
+#[derive(Debug, Clone)]
+pub struct Client {
+    /// The base URL this client sends requests to
+    pub(crate) base_url: String,
+    /// HTTP client to interact with Pasqal Cloud service
+    pub(crate) client: reqwest_middleware::ClientWithMiddleware,
+    pub(crate) project_id: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Response<T> {
+    pub data: T,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct GetDeviceResponseData {
+    pub status: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct GetDeviceSpecsResponseData {
+    pub specs: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct CreateBatchResponseData {
+    pub id: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct GetBatchResponseData {
+    pub status: BatchStatus,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct CancelBatchResponseData {}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct GetBatchResultLinksResponseData {
+    pub results_links: HashMap<String, String>,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct Job {
+    pub runs: i32,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct Batch {
+    pub sequence_builder: String,
+    pub jobs: Vec<Job>,
+    pub device_type: String,
+    pub project_id: String,
+}
+
+impl Client {
+    pub async fn get_device(
+        &self,
+        device_type: DeviceType,
+    ) -> Result<Response<GetDeviceResponseData>> {
+        let url = format!(
+            "{}/core-fast/api/v1/devices?device_type={}",
+            self.base_url, device_type,
+        );
+        self.get(&url).await
+    }
+    /// Pasqal Cloud works with batches of jobs rather than
+    /// individual jobs, see:
+    /// https://docs.pasqal.com/cloud/batches/
+    pub async fn create_batch(
+        &self,
+        sequence: String,
+        job_runs: i32,
+        device_type: DeviceType,
+    ) -> Result<Response<CreateBatchResponseData>> {
+        let url = format!("{}/core-fast/api/v1/batches", self.base_url);
+        let batch = Batch {
+            sequence_builder: sequence,
+            jobs: Vec::from([Job { runs: job_runs }]),
+            device_type: device_type.to_string(),
```

Comment:
I don't understand why, but in Python after the rebase this failed to marshall into the correct object. The request was sent using "EMUFREE" rather than "EMU_FREE" as it should be. 

I didn't bother fixing it because I don't think I want to keep the names of the backend in the code, so I will rework this later to use `String` throughout. But in the spank plugin pr.